### DISCUSSION TITLE
Allow augmenting AppInitialProps

### DISCUSSION THREAD

**Discussion by SleepWalker** [vercel/next.js#21424] [Ideas]
**Describe the feature you'd like to request**

Currently it is not that easy to augment `AppInitialProps` with additional props. That means that you can not provide more strict types, when using custom `_app`.

If you use the following snippet in `_app.tsx`:

```
declare module 'next/app' {
  interface AppInitialProps {
    theme: Client['theme'];
  }
}
```

It won't affect the return type of `App.getInitialProps()` and will completely override `AppInitialProps` that are merged in `AppProps` (so you'll need to manually declare `pageProps` to fix an issue).

Here is demo: https://codesandbox.io/s/goofy-architecture-0wdxk?file=/src/demo1.tsx. There are two commented lines, that should be uncommented in order to fix errors.

In order to properly augment types, you'll need to augment `next/dist/next-server/lib/utils` as follows:

```js
declare module 'next/dist/next-server/lib/utils' {
  interface AppInitialProps {
    theme: Client['theme'];
  }
}
```

But this will fail, because `AppInitialProps` is declared as `type` (not `interface`) inside `utils` module.

**Describe the solution you'd like**

In order to allow correct module augmentation the `AppInitialProps`  should be declared as interface in `next/dist/next-server/lib/utils`. This fix should also allow extending `AppInitialProps` by augmenting `next/app` module.

**Describe alternatives you've considered**

I think, that original solution is the most suitable one

**Comment by SleepWalker** [comment]
@timneutkens Hello, can you explain why this was moved into discussion? This issue can be fixed in one line. I can submit a PR if it will be accepted. What should be discussed and who can approve the change?

**Comment by luixo** [comment]
@timneutkens @SleepWalker 
Let me break down year-long silence on this topic.
I see no other way to make strong types on custom `App` without rewriting some of the already existing types manually.

**Reply by SleepWalker** [reply]
Hmmm. Actually in my current codebase I'm using module augmentation and it works as expected:

```ts
declare module 'next/app' {
  interface AppInitialProps {
    pageProps: any;
    site: SiteProps | null;
    locale: string;
    location: Location;
  }
}
```

**Reply by luixo** [reply]
That's actually true for me as well.
Seems like this discussion is resolved.
